language/return_spec.rb

```diff
@@ -249,4 +249,218 @@ def f
+            ScratchPad << "before if"
+            if true
+              return
+            end
+
+            ScratchPad << "after if"
+          END_OF_CODE
+
+          load @filename
+          ScratchPad.recorded.should == ["before if"]
+        end
+      end
+
+      describe "within while loop" do
+        it "is allowed" do
+          File.write(@filename, <<-END_OF_CODE)
+            ScratchPad << "before while"
+            while true
+              return
+            end
+
+            ScratchPad << "after while"
+          END_OF_CODE
+
+          load @filename
+          ScratchPad.recorded.should == ["before while"]
+        end
+      end
+
+      describe "within a begin" do
+        it "is allowed in begin block" do
+          File.write(@filename, <<-END_OF_CODE)
+            ScratchPad << "before begin"
+            begin
+              return
+            end
+
+            ScratchPad << "after begin"
+          END_OF_CODE
+
+          load @filename
+          ScratchPad.recorded.should == ["before begin"]
+        end
+
+        it "is allowed in ensure block" do
+          File.write(@filename, <<-END_OF_CODE)
+            ScratchPad << "before begin"
+            begin
+            ensure
+              return
+            end
+
+            ScratchPad << "after begin"
+          END_OF_CODE
+
+          load @filename
+          ScratchPad.recorded.should == ["before begin"]
+        end
+
+        it "is allowed in rescue block" do
+          File.write(@filename, <<-END_OF_CODE)
+            ScratchPad << "before begin"
+            begin
+              raise
+            rescue RuntimeError
+              return
+            end
+
+            ScratchPad << "after begin"
+          END_OF_CODE
+
+          load @filename
+          ScratchPad.recorded.should == ["before begin"]
+        end
+
+        it "fires ensure block before returning" do
+          ruby_exe(<<-END_OF_CODE).should == "within ensure\n"
+            begin
+              return
+            ensure
+              puts "within ensure"
+            end
+
+            puts "after begin"
+          END_OF_CODE
+        end
+
+        it "does not fire ensure block before returning while loads file" do
+          File.write(@filename, <<-END_OF_CODE)
+            ScratchPad << "before begin"
+            begin
+              return
+            ensure
+              ScratchPad << "within ensure"
+            end
+
+            ScratchPad << "after begin"
+          END_OF_CODE
+
+          load @filename
+          ScratchPad.recorded.should == ["before begin"]
+        end
+
+        it "swallows exception if returns in ensure block" do
+          File.write(@filename, <<-END_OF_CODE)
+            begin
+              raise
+            ensure
+              ScratchPad << "before return"
+              return
+            end
+          END_OF_CODE
+
+          ->() {
+            load @filename
+          }.should_not raise_error
```

Comment:
Minor, but `should_not raise_error` is not so useful, leaving outside the lambda would be clear it's expected to not raise an error. I'll remove it.
I actually want to remove all `should_not raise_error`, because positive expectations are much more precise, but there are quite a few of them.

Reply:
`should_not raise_error` expresses the test case intention here. If feature is broken and exception propagates with such `raise_error` expectation test will fail. Without `raise_error` expectation test will considered as just broken/outdated.

It's a minor issue and if you consider `should_not raise_error` expectation as excessive - I will remove it.

Reply:
Yes, it makes a difference between an `error` and `failure` in mspec.
I personally see no value to that difference. I see it as either the spec passes or it does not.
If the spec fails because of a method returning a different value or because it raises an exception it's the same to me, it doesn't pass and needs fixing :)
If the spec raises an exception, the interpreter is behaving wrong whether it's an "error" or "failure" (or the spec is broken of course).